test/fhirtypes/StructureDefinition.test.ts

```diff
@@ -97,11 +89,130 @@ describe('StructureDefinition', () => {
+      expect(undefinedEl).toBeUndefined();
+    });
+
+    // References
+    it('should find a reference choice by path', () => {
+      const basedOnNoChoice = observation.findElementByPath('basedOn', getResolver(defs));
+      const basedOnChoice = observation.findElementByPath(
+        'basedOn[MedicationRequest]',
+        getResolver(defs)
+      );
+      expect(basedOnChoice).toBeDefined();
+      expect(basedOnChoice.id).toBe('Observation.basedOn');
+      expect(basedOnChoice).toBe(basedOnNoChoice);
+    });
+
+    it('should not find an incorrect reference choice by path', () => {
+      const basedOn = observation.findElementByPath('basedOn[foo]', getResolver(defs));
+      expect(basedOn).toBeUndefined();
+    });
+
+    // Slicing
+    it('should find a sliced element by path', () => {
+      const VSCat = respRate.findElementByPath('category[VSCat]', getResolver(defs));
+      expect(VSCat).toBeDefined();
+      expect(VSCat.id).toBe('Observation.category:VSCat');
+    });
+
+    it('should find a child of a sliced element by path', () => {
+      const VSCatID = respRate.findElementByPath('category[VSCat].id', getResolver(defs));
+      expect(VSCatID).toBeDefined();
+      expect(VSCatID.id).toBe('Observation.category:VSCat.id');
+    });
+
+    it('should find a re-sliced element by path', () => {
+      const jsonReslice = JSON.parse(
+        fs.readFileSync(
+          path.join(__dirname, '../fhirdefs/testdefs/patient-telecom-reslice-profile.json'),
+          'utf-8'
+        )
+      );
+      const reslice = StructureDefinition.fromJSON(jsonReslice);
+      const emailWorkEmail = reslice.findElementByPath('telecom[email][workEmail]');
+      expect(emailWorkEmail).toBeDefined();
+      expect(emailWorkEmail.sliceName).toBe('email/workEmail');
+    });
+
+    // Choices
+    it('should make explicit a non-existent choice element by path', () => {
+      const originalLength = observation.elements.length;
+      const valueX = observation.findElementByPath('value[x]');
+      expect(valueX.slicing).toBeUndefined();
+      const valueQuantity = observation.findElementByPath('valueQuantity', getResolver(defs));
+      expect(valueQuantity).toBeDefined();
+      expect(valueQuantity.id).toBe('Observation.value[x]:valueQuantity');
+      expect(valueX.slicing).toBeDefined();
+      expect(observation.elements.length).toBe(originalLength + 1);
+    });
+
+    it('should make explicit a non-existent choice element by child path', () => {
+      const originalLength = observation.elements.length;
+      const valueX = observation.findElementByPath('value[x]');
+      expect(valueX.slicing).toBeUndefined();
+      const valueQuantity = observation.findElementByPath(
+        'valueQuantity.system',
+        getResolver(defs)
+      );
+      expect(valueQuantity).toBeDefined();
+      expect(valueQuantity.id).toBe('Observation.value[x]:valueQuantity.system');
+      expect(valueX.slicing).toBeDefined();
```

Comment:
To be thorough, I guess we can/should ensure the discriminator values are correct here too.